sync/src/relayer/compact_block_process.rs

```diff
@@ -273,3 +156,182 @@ impl<'a> HeaderProvider for CompactBlockMedianTimeView<'a> {
+        let header_view = {
+            let total_difficulty = parent.total_difficulty() + header.difficulty();
+            crate::types::HeaderView::new(header, total_difficulty)
+        };
+
+        let state = shared.state().peers();
+        state.may_set_best_known_header(peer, header_view);
+
+        return StatusCode::CompactBlockAlreadyStored.with_context(block_hash);
+    } else if status.contains(BlockStatus::BLOCK_RECEIVED) {
+        // block already in orphan pool
+        return Status::ignored();
+    } else if status.contains(BlockStatus::BLOCK_INVALID) {
+        return StatusCode::BlockIsInvalid.with_context(block_hash);
+    }
+
+    let store_first = tip.number() + 1 >= header.number();
+    let parent = shared.get_header_view(&header.data().raw().parent_hash(), Some(store_first));
+    if parent.is_none() {
+        debug_target!(
+            crate::LOG_TARGET_RELAY,
+            "UnknownParent: {}, send_getheaders_to_peer({})",
+            block_hash,
+            peer
+        );
+        active_chain.send_getheaders_to_peer(nc.as_ref(), peer, &tip);
+        return StatusCode::CompactBlockRequiresParent.with_context(format!(
+            "{} parent: {}",
+            block_hash,
+            header.data().raw().parent_hash(),
+        ));
+    }
+
+    // compact block is in pending
+    let pending_compact_blocks = shared.state().pending_compact_blocks();
+    if pending_compact_blocks
+        .get(&block_hash)
+        .map(|(_, peers_map, _)| peers_map.contains_key(&peer))
+        .unwrap_or(false)
+    {
+        return StatusCode::CompactBlockIsAlreadyPending.with_context(block_hash);
+    }
+
+    // compact header verification
+    let fn_get_pending_header = {
+        |block_hash| {
+            pending_compact_blocks
+                .get(&block_hash)
+                .map(|(compact_block, _, _)| compact_block.header().into_view())
+                .or_else(|| {
+                    shared
+                        .get_header_view(&block_hash, None)
+                        .map(|header_view| header_view.into_inner())
+                })
+        }
+    };
+    let median_time_context = CompactBlockMedianTimeView {
+        fn_get_pending_header: Box::new(fn_get_pending_header),
+    };
+    let header_verifier = HeaderVerifier::new(&median_time_context, shared.consensus());
+    if let Err(err) = header_verifier.verify(&header) {
+        if err
+            .downcast_ref::<HeaderError>()
+            .map(|e| e.is_too_new())
+            .unwrap_or(false)
+        {
+            return Status::ignored();
+        } else {
+            shared
+                .state()
+                .insert_block_status(block_hash.clone(), BlockStatus::BLOCK_INVALID);
+            return StatusCode::CompactBlockHasInvalidHeader
+                .with_context(format!("{} {}", block_hash, err));
+        }
+    }
+
+    Status::ok()
+}
+
+type PendingCompactBlockMap = HashMap<
```

Comment:
Do not repeat type alias it, please refer to the original definition

Reply:
import it, done